feat(vim-patch.sh): add n flag to list possible N/A patches

janlazo opened the pull request:
Scan all unmerged Vim commits to determine which commits are N/A for Neovim.
vim_na_regexp.txt contains the file patterns to filter out N/A files in a Vim commit.
If version.c remains and it has the expected 2-liner update, then it's considered N/A.

------

vim_na_files.txt contains "absolute filepath" of N/A files.
It relies on vim_na_regexp.txt to reduce file size.

https://github.com/neovim/neovim/discussions/36326

<!--
  Thank you for contributing to Neovim!
  If this is your first time, check out https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#pull-requests-prs
  for our PR guidelines.
-->

janlazo committed:
feat(vim-patch.sh): add n flag to list possible N/A patches

Scan all unmerged Vim commits to determine which commits are N/A
for Neovim.
vim_na_regexp.txt contains the file patterns to filter out N/A files
in a Vim commit.
If version.c remains and it has the expected 2-liner update,
then it's considered N/A.

TODO:
Find N/A patches based on exact filepath blacklist.
It should depend on the regexp blacklist to reduce file size.

https://github.com/neovim/neovim/discussions/36326

janlazo committed:
feat(vim-patch.sh): detect possible N/A patches from file blacklist

vim_na_files.txt contains "absolute filepath" of N/A files.
It relies on vim_na_regexp.txt to reduce file size.

Replace temorary file with process substitution and local variable.
Assume ARGS_MAX is never reached for applicable files in Vim commit.

janlazo committed:
fix: 'vim-patch.sh -n' prints tokens

janlazo committed:
fix(vim-patch.sh): no xargs, use grep -f

BSD xargs and GNU xargs differ so refactor it to bash while-read loop.

"grep -f" is faster than parsing a file for inlined patterns via -e.

zeertzjq commented:
Hmm, I tested it again. It runs for a bit less than 2 minutes, but then it exits with status 123 and still doesn't output anything...

zeertzjq commented on scripts/vim-patch.sh:
```diff
@@ -886,7 +887,36 @@ review_pr() {
-while getopts "hlLmMVp:P:g:r:s" opt; do
+_check_patch() {
+  patch=$1
+  cd "$VIM_SOURCE_DIR_DEFAULT"
```
```suggestion
  cd "${VIM_SOURCE_DIR}"
```

janlazo committed:
fix(vim-patch.sh): is_na_patch()

is_na_patch() succeeds if Vim patch is N/A.
Keeps working directory via "git -C".
Format vim-patch: token outside is_na_patch().

Tolerate diff format across git versions
since version.c separator between patch numbers can be top/below
the patch number.

janlazo commented on scripts/vim-patch.sh:
```diff
@@ -886,7 +887,41 @@ review_pr() {
 }
 
-while getopts "hlLmMVp:P:g:r:s" opt; do
+is_na_patch() {
```
Can add a new flag to only check 1 patch
but it may have more value in its own function to return the list of relevant files.

janlazo committed:
fix(vim-patch): whitespace typo

Co-authored-by: zeertzjq <zeertzjq@outlook.com>

zeertzjq commented:
I'm not sure if I understand the question. However I have some specific examples about preprocessing:
- If a patch's changes are limited to textprop.c and test_textprop.vim (both of which are covered by vim_na_regexp.txt), it's mostly likely that the entire patch is N/A. Example: [patch 9.1.1401](https://github.com/vim/vim/commit/56b1207e66d8db82dabf3f36aadf24caea61aa71).
- If a patch changes both drawline.c and test_textprop.vim, even though the test case in test_textprop.vim still cannot be used directly in Nvim, it may be a reference for a Lua functional test, so in this case the test_textprop.vim shouldn't be filtered out when preprocessing. Example: [patch 9.1.1798](https://github.com/vim/vim/commit/d9318acc0223bafd3ed5e7f8cdde26328bf200e7).

jamessan commented on scripts/vim_na_regexp.txt:
```diff
@@ -0,0 +1,90 @@
+^src/gui[_.]
+^src/if_
+^src/install
+^src/libvterm/
```
@janlazo Why was libvterm excluded? We vendored it, so wouldn't the Vim patches be something that should be evaluated?

janlazo replied:
To my knowledge, Vim has to modify `libvterm` in C89 to be compiled with Vim compared to Neovim. Meaning, certain Vim patches for C89 compatibility shouldn't be applicable (normally). That C89 requirement should be loosened by now but I don't know when (ex. v9.1.x?).

Since Vim's `src/terminal.c` and related  files for implementing and testing `:terminal` are also marked as N/A, then it's possible that valid patches for `libvterm` are wrongly marked as N/A for Neovim **IFF** there are no applicable files remaining. Some `src/*.c`,  `runtime/doc/*.txt` remain relevant to prevent these patches from being marked as N/A.

@jamessan Do you know a specific patch that shouldn't be N/A from a `version.c` PR since https://github.com/neovim/neovim/commit/e15f79c659f8895afdb946333757ad5517329805 and https://github.com/neovim/neovim/pull/36895 ? I see 5 `libvterm` patches but there could be more.

jamessan replied:
v9.2.0279, v9.2.0288, and v9.2.0297 are the main ones since they're bug fixes.

janlazo replied:
b346b87fb25779f2d9803fae0e378595611f79fb this one?

We can "revert" them via https://github.com/neovim/neovim/blob/master/scripts/vimpatch_token_reverts.txt . To prevent in the future. whitelist either the source or the test file or both?

https://github.com/neovim/neovim/pull/30011 v0.3.3. so b00df7aa388994119346a21d77b0d0db2a0a5e9f patch 9.0.1684.

```
54ffef7eb patch 9.2.0569: out-of-bounds access in libvterm CSI 8 t resize
e29f33ef5 patch 9.2.0336: libvterm: no terminal reflow support
8d23fcb60 patch 9.2.0315: missing bound-checks
77e7a40af patch 9.2.0297: libvterm: can improve CSI overflow code
71a0a552c patch 9.2.0288: libvterm: signed integer overflow parsing long CSI args
fe05143f5 patch 9.2.0279: terminal: out-of-bounds write with overlong CSI argument list
d13f13529 patch 9.2.0264: Cannot disable kitty keyboard protocol in vim :terminal
b24a1d929 patch 9.1.2134: Terminal doesn't handle split UTF-8 sequence after ASCII
cde8ff63e patch 9.1.1078: Terminal ansi colors off by one after tgc reset
ec8deb092 patch 9.0.1940: wrong upstream version in libvterm README
ceee7a808 patch 9.0.1927: patch 1916 (fixed terminal size) not optimal
aa64ba158 patch 9.0.1916: Crash when allocating large terminal screen
```

specific to c,h files

```
54ffef7eb patch 9.2.0569: out-of-bounds access in libvterm CSI 8 t resize
e29f33ef5 patch 9.2.0336: libvterm: no terminal reflow support
8d23fcb60 patch 9.2.0315: missing bound-checks
77e7a40af patch 9.2.0297: libvterm: can improve CSI overflow code
71a0a552c patch 9.2.0288: libvterm: signed integer overflow parsing long CSI args
fe05143f5 patch 9.2.0279: terminal: out-of-bounds write with overlong CSI argument list
d13f13529 patch 9.2.0264: Cannot disable kitty keyboard protocol in vim :terminal
b24a1d929 patch 9.1.2134: Terminal doesn't handle split UTF-8 sequence after ASCII
cde8ff63e patch 9.1.1078: Terminal ansi colors off by one after tgc reset
ceee7a808 patch 9.0.1927: patch 1916 (fixed terminal size) not optimal
aa64ba158 patch 9.0.1916: Crash when allocating large terminal screen
```
